knimin/lib/data_access.py

```diff
@@ -2530,19 +2554,168 @@ def delete_study(self, study_id):
-                raise ValueError('Deletion of study %s failed.' % study_id)
-            cur.execute('COMMIT')
+            TRN.add(sql, [study_id])
+
+    def _sample_exists(self, sample_id):
+        """Checks whether a sample ID exists
+
+        Parameters
+        ----------
+        sample_id : str
+
+        Returns
+        ------
+        bool
+        """
+        sql = """SELECT EXISTS (SELECT 1 FROM pm.sample
+                                WHERE sample_id = %s)"""
+        return self._con.execute_fetchone(sql, [sample_id])[0]
+
+    def _barcode_exists(self, barcode):
+        """Confirms that a barcode exists
+
+        Parameters
+        ----------
+        barcode : str
+
+        Raises
+        ------
+        ValueError
+            If the barcode does not exist
+        """
+        sql = """SELECT EXISTS (SELECT 1 FROM barcodes.barcode
+                                WHERE barcode = %s)"""
+        if not self._con.execute_fetchone(sql, [barcode])[0]:
+            raise ValueError('Barcode %s does not exist.' % barcode)
+
+    def create_samples(self, samples):
+        """Creates samples
+
+        Parameters
+        ----------
+        samples : list of dict
+            {
+             id : str (mandatory),
+             is_blank : bool (optional, default: False),
+             barcode : str (optional),
+             notes : str (optional)
+            }
+
+        Raises
+        ------
+        ValueError
+            If there is error creating the samples
+        """
+        with TRN:
+            for sample in samples:
+                sample_id = sample['id']
+                if self._sample_exists(sample_id):
+                    raise ValueError('Sample ID %s already exists.'
+                                     % sample_id)
+                barcode = sample.get('barcode') or None
+                if barcode:
+                    self._barcode_exists(barcode)
+                is_blank = sample.get('is_blank') or False
+                notes = sample.get('notes') or None
+                sql = """INSERT INTO pm.sample (sample_id, is_blank, barcode,
+                                                notes)
+                         VALUES (%s, %s, %s, %s)
+                         RETURNING sample_id"""
+                TRN.add(sql, [sample_id, is_blank, barcode, notes])
+
+    def edit_samples(self, samples):
+        """Edits properties of existing samples
+
+        Parameters
+        ----------
+        samples : list of dict
+            {
+             id : str,
+                ID of the sample to edit
+             is_blank : bool (optional, default: False),
+             barcode : str (optional),
+             notes : str (optional)
+            }
+
+        Raises
+        ------
+        ValueError
+            If there is error editing the samples' properties
+        """
+        with TRN:
+            for sample in samples:
+                sample_id = sample['id']
```

Comment:
I see that they are duplicated. I couldn't figure out an easy way to move them into a separate function and to reduce lines of codes. Originally I had everything (create/edit/delete) in one single function. @antgonza indicated (?) that it may be a better idea to create classes like "study", "sample", "plate". I like this idea. For now maybe let me continue with what I did so that we can save some time.